client/consensus/rhd/src/lib.rs

```diff
@@ -1,253 +1,358 @@
+            match self.cm_rx.unwrap().poll()? {
+                Async::Ready(Some(commit_msg)) => {
+                    // stuff to do
+                    self.mb_tx.unbounded_send(msg);
+                }
+                _ => println!(""),
+            }
+        }
+
+        // poll agreement and send to mb_tx channel
+        // XXX: check agreement poll ability
+        // match self.agreement.poll()? {
+        //     Async::Ready(Some(msg)) => {
+        //	// stuff to do
+        //	// the result of poll of agreement is Committed<>, deal with it
+        //	self.mb_tx.unbounded_send(msg);
+
+        //     },
+        //     _ => {}
+        // }
+
+        Ok(Async::NotReady)
     }
 }
 
-
-impl<B> RhdWorker<B> where
+impl<B> RhdWorker<B>
+where
     B: BlockT + Clone + Eq,
     B::Hash: ::std::hash::Hash,
 {
     pub fn new(
-	authorities: Vec<AuthorityID>,  // needed?
-	tc_rx,
-	ts_tx,
-	mb_tx,
-	ib_rx,
+        authorities: Vec<AuthorityID>, // needed?
+        tc_rx: UnboundedReceiver,
+        ts_tx: UnboundedSender,
+        mb_tx: UnboundedSender,
+        ib_rx: UnboundedReceiver,
     ) -> RhdWorker {
+        // let rhd_context = RhdContext {
+        //     // parent_hash: hash.clone(),
+        //     // cache: self.round_cache.clone(),
+        //     // round_timeout_multiplier: self.round_timeout_multiplier,
+        //     // key: self.key.clone(),
+        //     authorities: authorities,
+        // };
+
+        // let (te_tx, te_rx) = mpsc::unbounded();
+        // let (fe_tx, fe_rx) = mpsc::unbounded();
+
+        // let mut agreement = rhododendron::agree(
+        //     rhd_context,
+        //     n,
+        //     max_faulty,
+        //     te_rx,  // input
+        //     fe_tx,  // output
+        // );
+
+        RhdWorker {
+            te_tx: None,
+            fe_rx: None,
+            cm_rx: None,
+            bn_tx: None,
+            tc_rx,
+            ts_tx,
+            mb_tx,
+            ib_rx,
+            current_round_block: None,
+            bft_task_running: false,
+            timer_running: false,
+        }
+    }
 
+    fn start_once_bft(&mut self) {
+        let arc_rhd_worker = Arc::new(self.clone());
+
+        let rhd_context = RhdContext {
+            authorities: authorities,
+            rhd_worker: arc_rhd_worker,
+        };
+
+        let (te_tx, te_rx) = mpsc::unbounded();
+        let (fe_tx, fe_rx) = mpsc::unbounded();
+        let (cm_tx, cm_rx) = mpsc::unbounded();
+        let (bn_tx, bn_rx) = mpsc::unbounded(); // block notify
+
+        let mut agreement = rhododendron::agree(
+            rhd_context,
+            n,
+            max_faulty,
+            te_rx, // input
+            fe_tx, // output
+        );
+
+        self.te_tx = Some(te_tx);
+        self.fe_rx = Some(fe_rx);
+        self.cm_rx = Some(cm_rx);
+        self.bn_tx = Some(bn_tx);
+        //self.agreement = Some(agreement);
+
+        tokio::spawn(futures::future::poll_fn(move || {
```

Comment:
Why not return an agreement directly so we do not need bn chanel

Reply:
Ok may be we need to reset chanells into None, but then we can just return the candidate